core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,891 @@
+use std::borrow::Cow;
+use std::rc::Rc;
+
+use bincode;
+use rocksdb::{self, SliceTransform, WriteBatch, DB};
+
+use ::*;
+use data::SizeOf;
+
+// Uniquely identifies an index for each base (the numeral index in PersistentState.indices).
+type IndexID = u32;
+
+// Incremented on each PersistentState initialization so that IndexSeq
+// can be used to create unique identifiers for rows.
+type IndexEpoch = u64;
+
+// Monotonically increasing sequence number since last IndexEpoch used to uniquely identify a row.
+type IndexSeq = u64;
+
+// RocksDB key used for storing meta information (like indices).
+const META_KEY: &'static [u8] = b"meta";
+
+struct PersistentIndex {
+    columns: Vec<usize>,
+    seq: IndexSeq,
+}
+
+// Store index information in RocksDB to avoid rebuilding indices on recovery.
+#[derive(Default, Serialize, Deserialize)]
+struct PersistentMeta {
+    indices: Vec<Vec<usize>>,
+    epoch: IndexEpoch,
+}
+
+impl PersistentIndex {
+    fn new(columns: Vec<usize>) -> Self {
+        Self { columns, seq: 0 }
+    }
+}
+
+/// PersistentState stores data in SQlite.
+pub struct PersistentState {
+    name: String,
+    db_opts: rocksdb::Options,
+    // We don't really want DB to be an option, but doing so lets us drop it manually in
+    // PersistenState's Drop by setting `self.db = None` - after which we can then discard the
+    // persisted files if we want to.
+    db: Option<DB>,
+    durability_mode: DurabilityMode,
+    indices: Vec<PersistentIndex>,
+    epoch: IndexEpoch,
+    primary_key: Option<Vec<usize>>,
+}
+
+impl SizeOf for PersistentState {
+    fn size_of(&self) -> u64 {
+        use std::mem::size_of;
+
+        size_of::<Self>() as u64
+    }
+
+    fn deep_size_of(&self) -> u64 {
+        self.size_of()
```

Comment:
This seems false? Though I guess we never want to explicitly evict from bases anyway? @ms705 ?